mypy/checker.py

```diff
@@ -2674,6 +2684,52 @@ def check_member_assignment(self, instance_type: Type, attribute_type: Type,
         infer = is_subtype(rvalue_type, get_type) and is_subtype(get_type, set_type)
         return rvalue_type if infer else set_type, get_type, infer
 
+    def check_parent_member_assignment(self, lvalue_type: Type,
+                                       lvalue_attribute_type: Type, rvalue_type: Type,
+                                       rvalue: Expression, context: MemberExpr) -> Type:
+        """Type inherited member assignment.
+
+        This defers to check_simple_assignment if the attribute is local to the instance_type,
+        otherwise it will inform the user where the attribute type was originally defined.
+
+        Return the inferred rvalue_type.
+        """
+        if isinstance(lvalue_type, (Instance)):
+            attribute_is_defined = lvalue_type.has_readable_member(context.name)
+            if isinstance(lvalue_type, UnionType):
+                direct_lvalue_vars = list(itertools.chain.from_iterable(
+                                          i.type.names for i in lvalue_type.items))
+                lvalue_bases = list(itertools.chain.from_iterable(
+                                    i.type.bases for i in lvalue_type.items))
+            else:
+                direct_lvalue_vars = lvalue_type.type.names
+                lvalue_bases = lvalue_type.type.bases
+                if '__metaclass__' in lvalue_type.type.names:
+                    # if the assigned metaclass is not class itself we have other problems
+                    metacls = lvalue_type.type.names['__metaclass__'].type.ret_type  # type: ignore
+                    lvalue_bases.append(metacls)
+                    attribute_is_defined = (attribute_is_defined or
+                                            metacls.has_readable_member(context.name))
+            if attribute_is_defined and context.name not in direct_lvalue_vars:
+                for base in lvalue_bases:
+                    if context.name in base.type.names:
+                        parent_context = base.type.names[context.name].node
+                        if isinstance(parent_context, Var) and parent_context.info.is_generic():
+                            # attribute may have been defined in this class but its type is
+                            # probably defined elsewhere, possibly in the instance itself.
+                            continue
+                        always_allow_any = (rvalue_type is not None and
+                                            not isinstance(rvalue_type, AnyType))
```

Comment:
This should be `lvalue_attribute_type` (in your notations) here and one line above. I understand you are trying to duplicate the logic in `check_simple_assignment()`. Maybe instead modify the latter to accept an optional extra message?